### PR TITLE
[DRAFT] Adds serverlist api

### PR DESCRIPTION
**Pull request by MaxIsJoe**
Still WIP, intended to replace the current serverlist api and move it over to centeral command.

- [X] Add model for what server data will be there.
- [X] Server heart beat in the form of a POST request.
- [X] Server data expires if there is no heart beat after 10 seconds.
- [ ] Add the ability for accounts to generate server tokens that can be requested over at `unitystation.org`
- [ ] Add the ability for admins and users to invalidate server tokens from the dashboard and `unitystation.org`
- [ ] Add the ability for admins to disable and ban token generation for users who are caught breaking Stationhub rules.
- [ ] Validate POST requests to see if they have a valid server token to prevent bad actors from spamming stationhub with fake servers.

### CODE REVIEW

File: src/serverlist/views.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from .models import Server
+
+
+def server_list(request):
+    servers = Server.objects.all()
+    data = []
+    ## TODO: ADD VALIDATION FOR APPROVED API KEYS SO THAT STRANGERS DONT SPAM STATIONHUB WITH INVALID SERVERS ##
+    for server in servers:
+        if server.is_expired():
+            server.delete()
```
Comment (Fogapod) [inline src/serverlist/views.py:17]: This seems expensive to delete each one separately
 Can it be done in 1 operation?
I also suspect race condition is present. Imagine 2 requests coming in at the same time. Both connections try deleting same server. will it error?

File: src/serverlist/views.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+@csrf_exempt
+def add_server(request):
+    if request.method != "POST":
+        return JsonResponse({"error": "Only POST requests are allowed"}, status=405)
+    try:
+        server_validation = validate_server_info(request)
+        if server_validation.status_code != 200:
+            return server_validation
+        data = json.loads(request.body)
```
Comment (Fogapod) [inline src/serverlist/views.py:39]: django doesn't have ability to create model from json? weird code

File: src/serverlist/views.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        linux_build = data.get("player_count")
+        mac_build = data.get("player_count")
+        build_version = data.get("player_count")
+        code_scan_version = data.get("player_count")
+
+        existing_server = Server.objects.filter(name=name).first()
+        if existing_server:
+            # In case the server updates its info quickly before the 10 seconds pass,
+            # we update everything while we're updating the experitation date as well
+            existing_server.ip_address = ip_address
```
Comment (Fogapod) [inline src/serverlist/views.py:56]: this is wrong because `.get()` can return None

File: src/serverlist/views.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            existing_server.image_link = image_link
+            existing_server.windows_build = windows_build
+            existing_server.linux_build = linux_build
+            existing_server.mac_build = mac_build
+            existing_server.build_version = build_version
+            existing_server.code_scan_version = code_scan_version
+            existing_server.updated_at = datetime.now()
+            existing_server.save()
+        else:
+            Server.objects.create(
```
Comment (Fogapod) [inline src/serverlist/views.py:69]: This will error because `.get()` can return None

File: src/serverlist/views.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                name=name,
+                ip_address=ip_address,
+                port=port,
+                player_count=player_count,
+                image_link=image_link,
+                updated_at=datetime.now(),
+            )
+        return JsonResponse({"success": "Server data added/updated successfully"}, status=200)
```
Comment (Fogapod) [inline src/serverlist/views.py:77]: you don't need this json body

File: src/serverlist/views.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    except Exception as e:
```
Comment (Fogapod) [inline src/serverlist/views.py:78]: enormous kilometer wide try block

File: src/serverlist/views.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    if (
+        windows_build is None
+        or linux_build is None
+        or mac_build is None
+        or build_version is None
+        or code_scan_version is None
+    ):
+        return JsonResponse({"error": "Missing build information."}, status=400)
+
+    return JsonResponse({"success": "Server info validated successfully"}, status=200)
```
Comment (Fogapod) [inline src/serverlist/views.py:110]: you don't need this json body

File: src/serverlist/models.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+from datetime import timedelta
+
+from django.db import models
+from django.utils import timezone
+
+
+class Server(models.Model):
+    name = models.CharField(max_length=100)
+    ip_address = models.CharField(max_length=50)
```
Comment (craftxbox) [inline src/serverlist/models.py:9]: Maximum length of a domain part is 63 characters, and an entire FQDN is 253. 50 could un-necessarily limit domains being used. 

For example, `cpee0dbd14eedcc-cme0dbd14eedca.cpe.net.cable.rogers.com` is a real hostname (I found on Google) that measures >50 chars.

**Comment by corp-0** [comment]
I started working this on #95

**Pull request closed without merging**